### ISSUE TITLE
Enable filepath.FilePath to change permissions of a file or directory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3124]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#3124|
|Type|enhancement|
|Created|2008-03-26 13:44:18Z|
|Branch|https://github.com/twisted/twisted/tree/fp-chmod-3124|

Enable chmodding.

Attachments:

* [filepath.chown.chmod.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ddc/ddca0f7ce8e8438c7b9d01aa7ac8309704b6b871/c762a4f5f5c1b15926cf10541f0ac66212dcde59.patch) (2764 bytes) - added by cyli on 2008-03-27 20:34:19Z - 
* [filepath.chown.chmod.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ddc/ddca0f7ce8e8438c7b9d01aa7ac8309704b6b871/506705aa4a920637a01be8c55b1e67adb078b60f.patch) (4604 bytes) - added by cyli on 2008-04-03 01:17:06Z - 
* [filepath.chown.chmod.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ddc/ddca0f7ce8e8438c7b9d01aa7ac8309704b6b871/25be5157ec592f1f576105c7a19ba1ad7fff1506.patch) (3548 bytes) - added by cyli on 2008-04-03 19:42:47Z - 
* [filepath.chown.chmod.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ddc/ddca0f7ce8e8438c7b9d01aa7ac8309704b6b871/dac2f26f5f0e2274aab7851768d5bbeb2bc0dd74.patch) (1660 bytes) - added by cyli on 2008-04-20 20:52:23Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3124 3124
type__enhancement enhancement
reporter__cyli cyli
priority__high high
milestone__ 
branch__branches_fp_chmod_3124 branches/fp-chmod-3124
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1206539058000000 1206539058000000
changetime__1208942225000000 1208942225000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli commented|
|-|-|

Hmm... I'm having problems trying to figure out how to test chown.  All I'm doing is calling os.chown, but I can't figure out how to write a test that would work on any machine.  If anyone has any ideas, please let me know.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli removed owner|
|-|-|

Ok, just mocked the os.chown and os.chmod functions for testing.  Here we go.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @cyli|
|-|-|

 * `TestCase` instances have a patch method that would be bit cleaner than your `Verification` class. See http://twistedmatrix.com/documents/current/api/twisted.trial.unittest.TestCase.html#patch
 * the two tests are mostly good, but I wonder if tests that actually call the underlying functions would not be useful too. Maybe calling chown with the current process id. You'd just have to skip it under windows.
 * maybe chown could do something a bit smarter, like taking string representing id. I'm thinking about that because of [#3127](https://github.com/twisted/twisted/issues/3127): mktap.getid basically does this, but it should be moved somewhere (and probably tested). Again, it's not blocking this ticket.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

Replying to [therve](#note_3):
  * _the two tests are mostly good, but I wonder if tests that actually call the underlying functions would not be useful too. Maybe calling chown with the current process id. You'd just have to skip it under windows._ This is what I was originally going to do with `FilePath.chmod`, but I wasn't sure it would work in windows.  I also do not really understand your suggestion for chown - if there is no user whose id corresponds with the process id, it would fail.  Do you mean, call it and make sure it fails?  Do you mean calling it with the current process's user id?  If so, how can you tell if it worked?  (presumably files were originally created with the owner and group being the current process's user id).
  * _maybe chown could do something a bit smarter, like taking string representing id. I'm thinking about that because of [#3127](https://github.com/twisted/twisted/issues/3127): mktap.getid basically does this, but it should be moved somewhere (and probably tested). Again, it's not blocking this ticket._ Looking at the documentation (http://twistedmatrix.com/documents/current/api/twisted.scripts.mktap.html#getid) and code (http://twistedmatrix.com/trac/browser/trunk/twisted/scripts/mktap.py#L26) for `mktap.getid`, it seems to take a uid and gid and probably return the string representation (username.groupname ?).  You seem to be suggesting that `FilePath.chown` does what `mktap.getid` does in reverse (take username.groupname and break it down to uid and gid, and then calling os.chown with the uid and gid).  I wouldn't mind doing so, and it wouldn't be difficult, but should the getid and resolveid (my made-up name for the reverse of getid) methods don't seem like they belong in `FilePath`.

Also, should chown just raise an error in Windows?

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

oops, that should read:  "I wouldn't mind doing so, and it wouldn't be difficult, getid and resolveid (my made-up name for the reverse of getid) methods don't seem like they belong in `FilePath`."

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli removed owner|
|-|-|

Alrighty, I added two additional tests: one that calls chmod, and one that calls chown.  I also had chown accept either uid/gid's or usernames/group names, and added a test for this as well.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @cyli|
|-|-|

 * test_chmod should actually uses 0555 and 0777 to work under windows (my first guess was wrong)
 * pwd and grp aren't available anywhere. I guess we should postpone this feature until we know where getid lands. Please remove this, I'll open a ticket eventually
 * you don't have to call restore manually after a patch, it will done automatically
 * test_chown_mockup should be named test_chownMockup (same for test_chmod_mockup)
 * test_chown_mockup can't work on windows, because of some limitations of the patch system. On the other hand, the feature will never work on Windows, so we'd better just skip this test too under windows.
 * test_chown docstring doesn't seem to be finished

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli removed owner|
|-|-|

Changes made, thanks.  :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [23195]) Branching to 'fp-chown-chmod-3124'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [23196]) Apply patch with some small changes.

Refs [#3124](https://github.com/twisted/twisted/issues/3124)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

`test_chmodMockup` seems pretty redundant with `test_chmod`.  I'd just keep `test_chmod`, since it really tests what the API is supposed to do.  `test_chmodMockup` is a whitebox test that doesn't seem to add any additional coverage.

_Almost_ contrariwise, `test_chown` is nearly redundant with `test_chownMockup`.  `test_chown` would succeed with a no-op implementation.  Of course, `test_chownMockup` would fail, so together they may demonstrate that the implementation is sound.  `test_chownMockup` is a bit whitebox, but since `chown` is not going to succeed if we try passing interesting arguments to it, that's probably a necessity.  It'd be nice if we had a more structured way of having verified fakes (which is basically what these two tests in combination are achieving), but maybe that's out of scope.  In any case, I don't see why `test_chownMockup` needs to be skipped on Windows.  It shouldn't fail there, should it? :)

It might be nice to have something in the `FilePath.chown` docstring indicating that it isn't going to work on Windows.  The `-1` arguments are really stupid too but I guess that's Python's fault.  I'm tempted to suggest we should use `None` here instead, but maybe diverging from the Python API is needless complexity (although I'm also tempted to suggest `changeMode` and `changeOwner` as method names too :).

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli commented|
|-|-|

Okay... do you want me to make these changes?  Or therve, since he's already made the branch?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @cyli|
|-|-|

Don't hesitate, I'll have no problem finding another ticket :).

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli removed owner|
|-|-|

As per discussion with exarkun, took out all chown stuff.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

Could you quickly explain why?

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

> Could you quickly explain why?

Sure.  There was some debate as to what chown's behavior should be in Windows, whether it should return a separate AttributeError to disambiguate between os.chown's error, and whether it should be in FilePath at all or whether FilePath should be split/subclassed/wrapped to provide Unix-specific functionality and Windows specific functionality.  At the end of the huge discussion, since I said I didn't actually need chown and I was just adding it to b complete/symmetric (because I needed chmod), everyone (the April 20th sprint) said that I should just take out the chown functionality until they can figure out what they want for FilePath.  It could always be reinstated in a future patch, but once it's in the API'd have to be supported forever (or for a very long time, at least).

Since chmod works in both Windows and Unix, it was fine to leave in.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve set owner to @therve|
|-|-|

Previous description was:
```
Enable chmodding and chowning.
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve commented|
|-|-|

(In [23397]) Branching to 'fp-chmod-3124'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set status to `closed`|
|-|-|

(In [23400]) Merge fp-chmod-3124

Author: cyli
Reviewers: therve, exarkun
Fixes [#3124](https://github.com/twisted/twisted/issues/3124)

Add the chmod method to FIlePath objects, allowing to change permissions.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
